### PR TITLE
Improved benchmarker

### PR DESCRIPTION
**Pull request by dotPiano**
@rdube this adds the functionalities that were missing in the first version.
- All measured values can be optionally stored for plotting.
- Beside timing measurements, also values can be stored, e.g. number of voxels created, number of matches, loop closure event, ...
- Timings and values are organized in hierarchical topics, making it easier to visually understand the statistics and to plot the factors contributing to a specific part of the code.
- Every collected measurement/value is associated with the timestamp and the ID of the update step, making easy to associate value for plots like segmentation time vs. number of new voxels.

Should be useful for you as well for papers/presentations, feel free to ask more details about how to use it :)

### CODE REVIEW

**Comment by rdube** [comment]
retest this please

**Comment by rdube** [comment]
retest this please

File: laser_slam/include/laser_slam/benchmarker.hpp [hunk 1]
```diff
@@ -2,128 +2,207 @@
...
+    // Sum of the values, needed for computing mean and standard deviations.
+    double sum_ = 0.0;
+
+    // Sum of squares of the values, needed for computing the standard deviation.
+    double sum_of_squares_ = 0.0;
 
-    /// \brief Sum of squares of the measurements, needed for computing the standard deviation.
-    double sum_of_squares_ { 0.0 };
+    // The number of values recorded.
+    uint64_t values_count_ = 0;
```
Comment (rdube) [inline laser_slam/include/laser_slam/benchmarker.hpp]: 0u

File: laser_slam/include/laser_slam/benchmarker.hpp [hunk 2]
```diff
@@ -2,128 +2,207 @@
...
 class ScopedTimer {
 
  public:
-
   /// \brief Default constructor. Starts the timer.
-  ScopedTimer(const std::string& name);
+  /// \param topic_name Name of the topic to which the measurement belongs.
+  ScopedTimer(const std::string& topic_name)
+   : topic_name_(topic_name)
+   , start_(Benchmarker::Clock::now()) {
```
Comment (rdube) [inline laser_slam/include/laser_slam/benchmarker.hpp]: bring , up?

File: laser_slam/src/benchmarker.cpp [hunk 3]
```diff
@@ -5,109 +5,219 @@
 #include <iomanip>
 #include <iostream>
 
+#include <boost/algorithm/string.hpp>
+#include <boost/filesystem.hpp>
```
Comment (rdube) [inline laser_slam/src/benchmarker.cpp]: We managed not to use boost so far. Would it be hard work to replace it with standard c++ functions?

File: laser_slam/include/laser_slam/benchmarker.hpp [hunk 4]
```diff
@@ -2,128 +2,207 @@
...
-  /// Measurements are sorted in alphabetical order.
-  static std::map<std::string, MeasurementStatistics_> statistics_;
+  // Map containing the values sorted in alphabetical order.
+  static std::map<std::string, ValueTopic> value_topics_;
+
+  // Starting time points of the measurements currently in progress.
+  static std::unordered_map<std::string, TimePoint> started_mesurements_;
+
+  // Timestamp of the step being currently benchmarked.
+  static TimePoint current_timestamp_;
```
Comment (rdube) [inline laser_slam/include/laser_slam/benchmarker.hpp]: I am not sure to grasp the need for "steps". I believe that when saving a measurement, the start time and the duration should be all which is required. We can chat about this tomorrow.

File: laser_slam_ros/include/laser_slam_ros/common.hpp [hunk 5]
```diff
@@ -2,6 +2,7 @@
 #define LASER_SLAM_ROS_COMMON_HPP_
 
 #include <laser_slam/parameters.hpp>
+#include <laser_slam/benchmarker.hpp>
```
Comment (rdube) [inline laser_slam_ros/include/laser_slam_ros/common.hpp]: alph

File: laser_slam/include/laser_slam/benchmarker.hpp (hunk 4 above)
Comment (dotPiano) [inline laser_slam/include/laser_slam/benchmarker.hpp]: Sure, we can wait until tomorrow!

File: laser_slam/src/benchmarker.cpp (hunk 3 above)
Comment (dotPiano) [inline laser_slam/src/benchmarker.cpp]: No problem, boost is more compact, but it's definitely possible to do the same without!

Comment (dotPiano) [inline laser_slam/src/benchmarker.cpp]: Actually not that easy. Standard C++ is unable to create directories. Let's discuss this tomorrow as well!

**Pull request merged**